### ISSUE TITLE
[BLOCKING] [python] DMatrix setting incorrect values from sliced array

### ISSUE THREAD

**Issue by adibsurani** [dmlc/xgboost#3841]
On xgboost 0.80 (tested on Windows with Python 2 and Python 3), creating a DMatrix from a sliced array will result in incorrect values being copied in some cases. Here is a minimal repro:
```python
import numpy as np
import xgboost as xgb

y = np.array([12, 34, 56], np.float32)[::2]
print(xgb.DMatrix([], y).get_label())  # prints [12, 34] instead of [12, 56]
print(xgb.DMatrix([], y + 0).get_label())  # prints [12, 56] as expected
```

This issue has popped up for LightGBM before (Microsoft/LightGBM#1205), for which their solution is to test `label.flags.c_contiguous` and make a copy of the array as necessary.

**Comment by hcho3** [comment]
@adibsurani Thanks, we'll fix it in the next release. For now, this issue will be listed as a "known issue."

**Comment by trivialfis** [comment]
Simply copying the whole data would results in higher peak memory usage.  Copying data in `c_api.cc` by reading array stride will require python's header to build XGBoost ...

**Comment by trivialfis** [comment]
@hcho3 I will look into this.
